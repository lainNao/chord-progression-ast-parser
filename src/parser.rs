use crate::lexer::Token;

/*

type ParsedResult = Section[]

type Section = {
    meta: SectionMeta[]
    codeBlocks: CodeBlock[]
}

type SectionMeta = {
    key: "section"
    value: string
} | {
    key: "repeat"
    value: number
}

type CodeBlock = CodeInfo[]     //CSVで別れたコード情報達

type CodeInfo = {
    meta: CodeInfoMeta[]
    codes: Code[]
}

type Code = string;
    ここ、場合によってはコードを以下に分割してもいいかも
        type Code = {
            raw: A#m7b5            //全部統合したやつ。以下は分解したやつ
            base: A~G
            accidental: #|b
            type: m|M|aug|dim|add|sus2|sus4
            extension: 2|3|b3|4|b5|-5|5|#5|b6|6|7|b9|9|#9|b11|11|#11|b13|13|#13
        }

type CodeInfoMeta = {
    key: "key"
    value: Cb | Cbm | C | Cm | C# | C#m | Db | Dbm | D | Dm | D# | D#m | Eb | Ebm | E | Em | Fb | Fbm | F | Fm | F# | F#m | Gb | Gbm | G | Gm | G# | G#m | Ab | Abm | A | Am | A# | A#m | Bb | Bbm | B | Bm
}

*/

#[derive(Debug, PartialEq, Clone)]
pub enum Ast {
    Section {
        meta: Option<Meta>,
        code_blocks: Vec<CodeBlock>,
    },
}

#[derive(Debug, PartialEq, Clone)]
pub struct Meta {
    key: String,
    value: String,
}

#[derive(Debug, PartialEq, Clone)]
pub struct CodeBlock {
    meta: Option<Meta>,
    codes: Vec<String>,
}

pub fn parser(tokens: &[Token]) -> Result<Ast, String> {
    let mut tokens = tokens.iter().peekable();
    let mut section_meta: Option<Meta> = None;
    let mut code_blocks: Vec<CodeBlock> = Vec::new();

    while let Some(token) = tokens.next() {
        match token {
            Token::MetaInfoStart => {
                if let Some(Token::MetaInfoKey(key)) = tokens.next() {
                    if let Some(Token::Equal) = tokens.next() {
                        if let Some(Token::Code(value)) = tokens.next() {
                            section_meta = Some(Meta {
                                key: key.clone(),
                                value: value.clone(),
                            });
                            #[cfg(debug_assertions)]
                            println!("Parser: Parsed section meta: {:?}", section_meta);
                        } else {
                            return Err(
                                "Error: Expected Token::Value after Token::Equal".to_string()
                            );
                        }
                    } else {
                        return Err(
                            "Error: Expected Token::Equal after Token::MetaInfo".to_string()
                        );
                    }
                } else {
                    return Err(
                        "Error: Expected Token::MetaInfo after Token::SectionMetaStart".to_string(),
                    );
                }
            }
            Token::CodeBlockSeparator => {
                let mut meta: Option<Meta> = None;
                let mut codes: Vec<String> = Vec::new();

                loop {
                    match tokens.peek() {
                        Some(Token::MetaInfoKey(key)) => {
                            tokens.next(); // Consume the peeked token
                            if let Some(Token::Equal) = tokens.next() {
                                if let Some(Token::Code(value)) = tokens.next() {
                                    meta = Some(Meta {
                                        key: key.clone(),
                                        value: value.clone(),
                                    });
                                    #[cfg(debug_assertions)]
                                    println!("Parser: Parsed code block meta: {:?}", meta);
                                } else {
                                    return Err("Error: Expected Token::Value after Token::Equal"
                                        .to_string());
                                }
                            } else {
                                return Err("Error: Expected Token::Equal after Token::MetaInfo"
                                    .to_string());
                            }
                        }
                        Some(Token::Code(code)) => {
                            tokens.next(); // Consume the peeked token
                            codes.push(code.clone());
                        }
                        Some(Token::CodeBlockSeparator) => {
                            tokens.next(); // Consume the peeked token
                            break;
                        }
                        _ => {
                            return Err("Unexpected token inside a code block".to_string());
                        }
                    }
                }

                code_blocks.push(CodeBlock { meta, codes });
            }
            _ => {}
        }
    }

    Ok(Ast::Section {
        meta: section_meta,
        code_blocks,
    })
}
