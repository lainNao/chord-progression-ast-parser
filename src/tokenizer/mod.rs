use crate::errors;

pub mod types;
pub mod util;

use types::{Token, ValueToken};
use util::is_token_char;

pub fn tokenize(input: &str) -> Result<Vec<Token>, String> {
    let mut tokens = Vec::new();
    let mut chars = input.chars().peekable();

    while let Some(ch) = chars.next() {
        match ch {
            '@' => tokens.push(Token::SectionMetaInfoStart),
            '(' => tokens.push(Token::MetaInfoStart),
            ')' => tokens.push(Token::MetaInfoEnd),
            '|' => tokens.push(Token::ChordBlockSeparator),
            '=' => tokens.push(Token::Equal),
            ',' => tokens.push(Token::Comma),
            '/' => tokens.push(Token::Slash),
            ' ' | 'ã€€' | '\t' => {}
            '\n' | '\r' => match tokens.last() {
                Some(Token::SectionMetaInfoKey(_)) => {
                    return Err(
                        errors::SECTION_META_INFO_KEY_SHOULD_NOT_CONTAINS_LINE_BREAK.to_string()
                    );
                }
                Some(Token::MetaInfoKey(_)) => {
                    return Err(errors::META_INFO_KEY_SHOULD_NOT_CONTAINS_LINE_BREAK.to_string());
                }
                Some(Token::MetaInfoValue(_)) => {
                    return Err(errors::META_INFO_VALUE_SHOULD_NOT_CONTAINS_LINE_BREAK.to_string());
                }
                Some(Token::Comma) => {
                    return Err(errors::CHORD_BLOCK_SHOULD_NOT_CONTAINS_LINE_BREAK.to_string());
                }
                _ => tokens.push(Token::LineBreak),
            },
            non_functional_char => {
                let mut token = String::new();
                token.push(non_functional_char);
                let second_token_from_last = if tokens.len() >= 2 {
                    tokens.get(tokens.len() - 2)
                } else {
                    None
                };

                // get token type
                let get_token_type_result = match tokens.last() {
                    Some(Token::SectionMetaInfoStart) => Ok(Some(ValueToken::SectionMetaInfoKey)),
                    Some(Token::MetaInfoStart) => Ok(Some(ValueToken::MetaInfoKey)),
                    Some(Token::Equal) => match second_token_from_last {
                        Some(Token::SectionMetaInfoKey(_)) => {
                            Ok(Some(ValueToken::SectionMetaInfoValue))
                        }
                        Some(Token::MetaInfoKey(_)) => Ok(Some(ValueToken::MetaInfoValue)),
                        _ => {
                            return Err([
                                errors::INVALID_TOKEN_TYPE.to_string(),
                                second_token_from_last.unwrap().to_string(),
                            ]
                            .join(": "));
                        }
                    },
                    Some(Token::Slash) => Ok(Some(ValueToken::Denominator)),
                    _ => Ok(Some(ValueToken::Chord)),
                };

                // if error, return
                let token_type = if let Ok(token_type) = get_token_type_result {
                    token_type
                } else {
                    return Err(get_token_type_result.unwrap_err());
                };

                // get token
                while let Some(&next_ch) = chars.peek() {
                    if is_token_char(next_ch) {
                        break;
                    }

                    token.push(next_ch);
                    chars.next();
                }

                // push token
                match token_type {
                    Some(ValueToken::SectionMetaInfoKey) => {
                        tokens.push(Token::SectionMetaInfoKey(token))
                    }
                    Some(ValueToken::SectionMetaInfoValue) => {
                        tokens.push(Token::SectionMetaInfoValue(token))
                    }
                    Some(ValueToken::MetaInfoKey) => tokens.push(Token::MetaInfoKey(token)),
                    Some(ValueToken::MetaInfoValue) => tokens.push(Token::MetaInfoValue(token)),
                    Some(ValueToken::Chord) => tokens.push(Token::Chord(token)),
                    Some(ValueToken::Denominator) => tokens.push(Token::Denominator(token)),
                    None => {
                        return Err([
                            errors::INVALID_TOKEN_TYPE.to_string(),
                            token_type.unwrap().to_string(),
                        ]
                        .join(": "));
                    }
                }
            }
        }
    }

    Ok(tokens)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[cfg(test)]
    mod success {
        use super::*;

        #[test]
        fn section_meta_info() {
            let input = "@section=A";
            let expected = vec![
                Token::SectionMetaInfoStart,
                Token::SectionMetaInfoKey("section".to_string()),
                Token::Equal,
                Token::SectionMetaInfoValue("A".to_string()),
            ];

            let lex_result = tokenize(input);
            println!("111 {:?}", lex_result);
            assert!(lex_result.is_ok());
            let tokens = lex_result.unwrap();
            assert_eq!(tokens, expected);
        }

        #[test]
        fn multiple_section_meta_info() {
            let input = "
                @section=A
                @sample=aaa
                ";
            let expected = vec![
                Token::LineBreak,
                Token::SectionMetaInfoStart,
                Token::SectionMetaInfoKey("section".to_string()),
                Token::Equal,
                Token::SectionMetaInfoValue("A".to_string()),
                Token::LineBreak,
                Token::SectionMetaInfoStart,
                Token::SectionMetaInfoKey("sample".to_string()),
                Token::Equal,
                Token::SectionMetaInfoValue("aaa".to_string()),
                Token::LineBreak,
            ];

            let lex_result = tokenize(input);
            assert!(lex_result.is_ok());
            let tokens = lex_result.unwrap();
            assert_eq!(tokens, expected);
        }

        #[test]
        fn chord_block() {
            let input = "
                |C|F|Fm|C|
                ";

            let expected = vec![
                Token::LineBreak,
                Token::ChordBlockSeparator,
                Token::Chord("C".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Fm".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("C".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
            ];

            let lex_result = tokenize(input);
            assert!(lex_result.is_ok());
            let tokens = lex_result.unwrap();
            assert_eq!(tokens, expected);
        }

        #[test]
        fn chord_block_with_fraction_chord() {
            let input = "
                |C|G/Bb|Am|Em/G|
                |F#m7-5/F#m7-5|Fbm13/G7|
                ";

            let expected = vec![
                Token::LineBreak,
                Token::ChordBlockSeparator,
                Token::Chord("C".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("G".to_string()),
                Token::Slash,
                Token::Denominator("Bb".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Am".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Em".to_string()),
                Token::Slash,
                Token::Denominator("G".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
                Token::ChordBlockSeparator,
                Token::Chord("F#m7-5".to_string()),
                Token::Slash,
                Token::Denominator("F#m7-5".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Fbm13".to_string()),
                Token::Slash,
                Token::Denominator("G7".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
            ];

            let lex_result = tokenize(input);
            assert!(lex_result.is_ok());
            let tokens = lex_result.unwrap();
            assert_eq!(tokens, expected);
        }

        #[test]
        fn chord_block_with_multiple_meta_info() {
            let input = "
                |(key=C)(sample=aaa)C|F|(key=Eb)Fm,(sample=bbb)Bb|C|
                ";

            let expected = vec![
                Token::LineBreak,
                Token::ChordBlockSeparator,
                Token::MetaInfoStart,
                Token::MetaInfoKey("key".to_string()),
                Token::Equal,
                Token::MetaInfoValue("C".to_string()),
                Token::MetaInfoEnd,
                Token::MetaInfoStart,
                Token::MetaInfoKey("sample".to_string()),
                Token::Equal,
                Token::MetaInfoValue("aaa".to_string()),
                Token::MetaInfoEnd,
                Token::Chord("C".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::MetaInfoStart,
                Token::MetaInfoKey("key".to_string()),
                Token::Equal,
                Token::MetaInfoValue("Eb".to_string()),
                Token::MetaInfoEnd,
                Token::Chord("Fm".to_string()),
                Token::Comma,
                Token::MetaInfoStart,
                Token::MetaInfoKey("sample".to_string()),
                Token::Equal,
                Token::MetaInfoValue("bbb".to_string()),
                Token::MetaInfoEnd,
                Token::Chord("Bb".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("C".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
            ];

            let lex_result = tokenize(input);
            println!("222 {:?}", lex_result);
            assert!(lex_result.is_ok());
            let tokens = lex_result.unwrap();
            assert_eq!(tokens, expected);
        }

        #[test]
        fn complicated() {
            let input = "
                @section=A
                @sample=aaa
                |C|C7|F|Fm7|
                |C|C7|F|Fm7|

                @section=B
                |(key=F)Gm|Gm|F|F|
                |Gm|Gm|F|F|
                ";

            let expected = vec![
                Token::LineBreak,
                // @section=A
                Token::SectionMetaInfoStart,
                Token::SectionMetaInfoKey("section".to_string()),
                Token::Equal,
                Token::SectionMetaInfoValue("A".to_string()),
                Token::LineBreak,
                // @sample=aaa
                Token::SectionMetaInfoStart,
                Token::SectionMetaInfoKey("sample".to_string()),
                Token::Equal,
                Token::SectionMetaInfoValue("aaa".to_string()),
                Token::LineBreak,
                // |C|C7|F|Fm7|
                Token::ChordBlockSeparator,
                Token::Chord("C".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("C7".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Fm7".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
                // |C|C7|F|Fm7|
                Token::ChordBlockSeparator,
                Token::Chord("C".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("C7".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Fm7".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
                Token::LineBreak,
                // @section=B
                Token::SectionMetaInfoStart,
                Token::SectionMetaInfoKey("section".to_string()),
                Token::Equal,
                Token::SectionMetaInfoValue("B".to_string()),
                Token::LineBreak,
                // |(key=F)Gm|Gm|F|F|
                Token::ChordBlockSeparator,
                Token::MetaInfoStart,
                Token::MetaInfoKey("key".to_string()),
                Token::Equal,
                Token::MetaInfoValue("F".to_string()),
                Token::MetaInfoEnd,
                Token::Chord("Gm".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Gm".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
                // |Gm|Gm|F|F|
                Token::ChordBlockSeparator,
                Token::Chord("Gm".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("Gm".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::Chord("F".to_string()),
                Token::ChordBlockSeparator,
                Token::LineBreak,
            ];

            let lex_result = tokenize(input);

            assert!(lex_result.is_ok());
            let tokens = lex_result.unwrap();
            assert_eq!(tokens, expected);
        }
    }

    #[cfg(test)]
    mod failed {
        use super::*;

        #[test]
        fn section_meta_info_key_should_not_contains_line_break() {
            let input = "
            @sect
            ion=A
            ";

            let lex_result = tokenize(input);
            assert!(lex_result.is_err());
            assert_eq!(
                lex_result.unwrap_err(),
                errors::SECTION_META_INFO_KEY_SHOULD_NOT_CONTAINS_LINE_BREAK
            );
        }

        #[test]
        fn meta_info_key_should_not_contains_line_break() {
            let input = "
                |(aaaa
                    aaaa=bbbb
                )C|
                ";

            let lex_result = tokenize(input);
            assert!(lex_result.is_err());
            assert_eq!(
                lex_result.unwrap_err(),
                errors::META_INFO_KEY_SHOULD_NOT_CONTAINS_LINE_BREAK
            );
        }

        #[test]
        fn meta_info_value_should_not_contains_line_break() {
            let input = "
                |(aaaa=bbbb
                    bbbb)C|
                ";

            let lex_result = tokenize(input);
            assert!(lex_result.is_err());
            assert_eq!(
                lex_result.unwrap_err(),
                errors::META_INFO_VALUE_SHOULD_NOT_CONTAINS_LINE_BREAK
            );
        }

        #[test]
        fn chord_block_should_not_contains_line_break() {
            let input = "
                |C,
                C7|F|Fm7|
                ";

            let lex_result = tokenize(input);
            println!("222 {:?}", lex_result);
            assert!(lex_result.is_err());
            assert_eq!(
                lex_result.unwrap_err(),
                errors::CHORD_BLOCK_SHOULD_NOT_CONTAINS_LINE_BREAK
            );
        }
    }
}
