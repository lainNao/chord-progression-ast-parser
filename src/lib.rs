mod errors;
mod parser;
mod tokenizer;

use parser::parse;
use serde_json::json;
use tokenizer::tokenize;
use wasm_bindgen::prelude::wasm_bindgen;

#[wasm_bindgen]
pub fn run(input: &str) -> Result<String, String> {
    let tokens = match tokenize(input) {
        Ok(t) => t,
        Err(e) => {
            return Err(format!("Tokenization Error: {}", e));
        }
    };
    let ast = match parse(&tokens) {
        Ok(ast) => ast,
        Err(e) => {
            return Err(format!("Parse Error: {}", e));
        }
    };

    let json = json!(ast);
    Ok(json.to_string())
}

#[cfg(test)]
mod tests {
    #[cfg(test)]
    mod success {
        use crate::{parser::parse, tokenizer::tokenize};
        use serde_json::json;

        #[test]
        fn complex_input_can_be_parsed() {
            let input: &str = "
            @section=Intro
            |[key=E]E|C#m(7)|Bm(7)|C#(7)|
            |F#m(7)|Am(7)|F#(7)|B|
            
            @section=Verse
            |E|C#m(7)|Bm(7)|C#(7)|
            |F#m(7)|Am(7)|F#(7)|B|
            
            @section=Chorus
            |[key=C]C|C(7)|FM(7)|Fm(7)|
            |C|C(7)|FM(7)|Dm(7)|
            |Em(7)|E(7)|
        
            @section=Interlude
            |C|A,B|

            |[key=C]C(M9)|CM(9)|
            ";

            let result = parse(&tokenize(input).unwrap());
            assert!(result.is_ok());
        }

        #[test]
        fn differ_major_9_vs_9_of_major() {
            let input: &str = "
            @section=Intro
            |[key=C]C(M9)|CM(9)|
            ";

            let result_json = json!(parse(&tokenize(input).unwrap()).unwrap());
            let expected = json!([
                {
                    "chord_blocks": [
                        [
                            {
                                "chord": {
                                    "Chord":{
                                        "detailed": {
                                            "accidental":null,
                                            "base":"C",
                                            "chord_type":"Major",
                                            "extensions": [
                                                "MajorNine"
                                            ]
                                        },
                                        "plain":"C(M9)"
                                    }
                                },
                                "denominator":null,
                                "meta_infos": [
                                    {
                                        "Key":{
                                            "value":"C_M"
                                        }
                                    }
                                ]
                            },
                            {
                                "chord": {
                                    "Chord":{
                                        "detailed": {
                                            "accidental":null,
                                            "base":"C",
                                            "chord_type":"Major",
                                            "extensions": [
                                                "Nine"
                                            ]
                                        },
                                        "plain":"CM(9)"
                                    }
                                },
                                "denominator":null,
                                "meta_infos": []
                            }
                        ]
                    ],
                    "meta_infos": [
                        {
                            "Section": {
                                "value":"Intro"
                            }
                        }
                    ]
                }
            ]);

            assert_eq!(result_json, expected);
        }
    }
}
