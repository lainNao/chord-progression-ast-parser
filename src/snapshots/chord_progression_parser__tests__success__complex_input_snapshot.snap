---
source: src/lib.rs
expression: parse_chord_progression_string(input)
---
Ok(
    [
        Section {
            meta_infos: [
                Section(
                    "Intro",
                ),
            ],
            chord_blocks: [
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [
                                Key(
                                    E_M,
                                ),
                            ],
                            chord_expression: Chord(
                                Chord {
                                    plain: "E",
                                    detailed: ChordDetailed {
                                        base: E,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C#m(7)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "Bm(7)",
                                    detailed: ChordDetailed {
                                        base: B,
                                        accidental: None,
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C#(7)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Br,
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "F#m(7)",
                                    detailed: ChordDetailed {
                                        base: F,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "Am(7)",
                                    detailed: ChordDetailed {
                                        base: A,
                                        accidental: None,
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "F#(7)",
                                    detailed: ChordDetailed {
                                        base: F,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "B",
                                    detailed: ChordDetailed {
                                        base: B,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
            ],
        },
        Section {
            meta_infos: [
                Section(
                    "Verse",
                ),
            ],
            chord_blocks: [
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "E",
                                    detailed: ChordDetailed {
                                        base: E,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C#m(7)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "Bm(7)",
                                    detailed: ChordDetailed {
                                        base: B,
                                        accidental: None,
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C#(7)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Br,
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "F#m(7)",
                                    detailed: ChordDetailed {
                                        base: F,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "Am(7)",
                                    detailed: ChordDetailed {
                                        base: A,
                                        accidental: None,
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "F#(7)",
                                    detailed: ChordDetailed {
                                        base: F,
                                        accidental: Some(
                                            Sharp,
                                        ),
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "B",
                                    detailed: ChordDetailed {
                                        base: B,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
            ],
        },
        Section {
            meta_infos: [
                Section(
                    "Chorus",
                ),
            ],
            chord_blocks: [
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [
                                Key(
                                    C_M,
                                ),
                            ],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C(7)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "FM(7)",
                                    detailed: ChordDetailed {
                                        base: F,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "Fm(7)",
                                    detailed: ChordDetailed {
                                        base: F,
                                        accidental: None,
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Br,
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C(7)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "FM(7)",
                                    detailed: ChordDetailed {
                                        base: F,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "Dm(7)",
                                    detailed: ChordDetailed {
                                        base: D,
                                        accidental: None,
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Br,
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "Em(7)",
                                    detailed: ChordDetailed {
                                        base: E,
                                        accidental: None,
                                        chord_type: Minor,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "E(7)",
                                    detailed: ChordDetailed {
                                        base: E,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [
                                            Seven,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
            ],
        },
        Section {
            meta_infos: [
                Section(
                    "Interlude",
                ),
            ],
            chord_blocks: [
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "A",
                                    detailed: ChordDetailed {
                                        base: A,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "B",
                                    detailed: ChordDetailed {
                                        base: B,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
            ],
        },
        Section {
            meta_infos: [],
            chord_blocks: [
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [
                                Key(
                                    C_M,
                                ),
                            ],
                            chord_expression: Chord(
                                Chord {
                                    plain: "C(M9)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [
                                            MajorNine,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
                Bar(
                    [
                        ChordInfo {
                            meta_infos: [],
                            chord_expression: Chord(
                                Chord {
                                    plain: "CM(9)",
                                    detailed: ChordDetailed {
                                        base: C,
                                        accidental: None,
                                        chord_type: Major,
                                        extensions: [
                                            Nine,
                                        ],
                                    },
                                },
                            ),
                            denominator: None,
                        },
                    ],
                ),
            ],
        },
    ],
)
